cmd/terramate/e2etests/exp_vendor_test.go

```diff
@@ -132,6 +135,104 @@ func TestVendorModuleRecursive1DependencyIsPatched(t *testing.T) {
 	assert.EqualStrings(t, "", string(got))
 }
 
+func TestModVendorRecursiveMustPatchAlreadyVendoredModules(t *testing.T) {
+	// This reproduces the bug below:
+	//   - ModA depends on ModZ
+	//   - ModB depends on ModZ
+	// executing:
+	//   - tm vendor download ModA <ref>
+	//   - tm vendor download ModB <ref>
+	// will patch the ModA files to use ../modules/ModZ/...
+	// but will not for ModB because ModZ is already vendored.
+
+	const filename = "main.tf"
+
+	setupModuleGit := func(name string, ref string, deps ...string) string {
+		moduleRepo := sandbox.New(t)
+		var mainContent string
+		for _, dep := range deps {
+			if mainContent != "" {
+				mainContent += "\n"
+			}
+
+			mainContent += Module(
+				Labels(name),
+				Str("source", dep+"?ref="+ref),
+			).String()
+		}
+
+		moduleRepo.RootEntry().CreateFile(filename, mainContent)
+		repoGit := moduleRepo.Git()
+		repoGit.CommitAll("add file")
+		repoGit.CheckoutNew("test")
+		repoGit.Push("main")
+		repoGit.Push("test")
+		return "git::file://" + moduleRepo.RootDir()
+	}
+
+	modZ := setupModuleGit("modZ", "main")
+	modA := setupModuleGit("modA", "main", modZ)
+	modB := setupModuleGit("modB", "main", modZ)
+
+	// modC is used to test for a different ref in the dependency. In this case,
+	// it must also vendor and patch.
+	modC := setupModuleGit("modC", "test", modZ)
+
+	// setup project
+	s := sandbox.NoGit(t)
+	s.RootEntry().CreateConfig(Terramate(
+		Config(),
+	).String())
+	tmcli := newCLI(t, s.RootDir())
+	res := tmcli.run("experimental", "vendor", "download", modA, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
+	res = tmcli.run("experimental", "vendor", "download", modB, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
+	res = tmcli.run("experimental", "vendor", "download", modC, "main")
+	assertRunResult(t, res, runExpected{IgnoreStdout: true})
```

Comment:
Here instead of running terramate we could just use modvendor.Vendor (and move the test to modvendor), the rest would remain the same I think.